_posts/2015-06-02-anatomy-of-service-objects-in-rails.md

```diff
@@ -0,0 +1,333 @@
+
+It's not uncommon to see code just instantiate the needed objects where needed, but this is not ideal.
+
+```ruby
+def process!(the_return,user)
+  if unpaid_items(the_return).any?
+    # Bad--this business logic is coupled to the 
+    #      creation of another service object
+    result = CheckoutService.new.charge!(unpaid_items(the_return))
+    unless result.charge_succeeded?
+      return Result.new(return_processed: false, 
+                        error_messages: result.error_messages)
+    end
+  end
+
+  # remainder of the method
+
+end
+```
+
+The problem is that `process!`—which should just be concerned with the details of processing a return—also has to know how to create a `CheckoutService` instance.  *This* means that if we need to change how `CheckoutService` is created, we have to change this method (and likely its tests).  We don't want this method to change unless the _business process_ of processing returns changes.
+
+The simplest thing to do is move it to a private method:
+
+```ruby
+class ReturnProcessor
+  def process!(the_return,user)
+    if unpaid_items(the_return).any?
+      # Good--our code just depends on an object 
+      #       that we can assume has been set up for us
+      result = checkout_service.charge!(unpaid_items(the_return))
+      unless result.charge_succeeded?
+        return Result.new(return_processed: false, 
+                          error_messages: result.error_messages)
+      end
+    end
+
+    # remainder of the method
+
+  end
+
+private
+
+  def checkout_service
+    @checkout_service ||= CheckoutService.new
+  end
+end
+```
+
+Although `ReturnProcessor` is still coupled to how `CheckoutService` gets created, the `process!` method no longer is.  Meaning it has less reasons to change, meaning our system is overall more resilient to changes.
+
+Another alternative is to allow callers to pass in a `CheckoutService` to `ReturnProcessor`'s constructor along with a sensible default.
+
+```ruby
+class ReturnProcessor
+
+  def initialize(checkout_service: nil)
+    @checkout_service = checkout_service || CheckoutService.new
+  end
+
+  def process!(the_return,user)
+    if unpaid_items(the_return).any?
+      result = checkout_service.charge!(unpaid_items(the_return))
+      unless result.charge_succeeded?
+        return Result.new(return_processed: false, 
+                          error_messages: result.error_messages)
+      end
+    end
+
+    # remainder of the method
+
+  end
+
+private
+
+  attr_reader :checkout_service
+end
+```
+
+This is more verbose, but is a useful pattern if:
+
+* The way `CheckoutService` is created is unstable and likely to change, and we wish to store that code outside this class
```

Comment:
Inconsistent lack of period at end of this bullet point (your other two bullets have it).